### ISSUE TITLE
Support for Indexed/Random Access Datasets in Megatron-Energon

### ISSUE THREAD

**Issue by leo1oel** [NVIDIA/Megatron-Energon#156]
While `shuffle_buffer_size` provides some randomness, it's still limited to buffered shuffling within the stream. For smaller datasets, I'd like to achieve true global shuffling with random access to any sample.

Does Megatron-Energon support indexed datasets that allow random access by index (similar to wids or other indexed formats)?

**Comment by philipp-fischer** [comment]
Hi @leo1oel,
note that typically, you train with multiple GPU ranks and each with multiple workers and each worker gets assigned random slice of each of many datasets. On top you have the shuffle buffer for each worker! That's a lot of randomness already.

We conveyed several analyses and found that this is sufficient in most cases in practice while it still provides an efficient access to disk. The impact on loading speed when doing fully random access is significant, since all layers in between (file systems, kernel, caches) are optimized for linear reading.

That said, it is possible to get fully random access although we do not recommend it at all. in `get_train_dataset`, simply set `max_samples_per_sequence=1`. It will be fully random but slower.
